File: app/lib/domain/models/task.dart
import 'status.dart';
import 'tag.dart';

class Task {
  final String id;
  final String title;
  final String? memo;
  final DateTime? dueDate;
  final int importance;
  final int urgency;
  final String? statusId;
  final List<String> tagIds;
  // Provider 層で解決されたオブジェクト
  final Status? status;
  final List<Tag> tags;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Task({
    required this.id,
    required this.title,
    this.memo,
    this.dueDate,
    this.importance = 1,
    this.urgency = 1,
    this.statusId,
    this.tagIds = const [],
    this.status,
    required this.tags,
    required this.createdAt,
    required this.updatedAt,
  });

  Task copyWith({
    String? id,
    String? title,
    String? memo,
    bool clearMemo = false,
    DateTime? dueDate,
    bool clearDueDate = false,
    int? importance,
    int? urgency,
    String? statusId,
    bool clearStatusId = false,
    List<String>? tagIds,
    Status? status,
    bool clearStatus = false,
    List<Tag>? tags,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Task(
      id: id ?? this.id,
      title: title ?? this.title,
      memo: clearMemo ? null : (memo ?? this.memo),
      dueDate: clearDueDate ? null : (dueDate ?? this.dueDate),
      importance: importance ?? this.importance,
      urgency: urgency ?? this.urgency,
      statusId: clearStatusId ? null : (statusId ?? this.statusId),
      tagIds: tagIds ?? this.tagIds,
      status: clearStatus ? null : (status ?? this.status),
      tags: tags ?? this.tags,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: api/internal/testfactory/task.go
package testfactory

import (
	"context"
	"testing"
	"time"

	"task-manager/api/internal/repository"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgtype"
	"github.com/jackc/pgx/v5/pgxpool"
)

type TaskParams struct {
	ID         uuid.UUID
	UserID     uuid.UUID
	Title      string
	StatusID   uuid.UUID
	Importance int32
	Urgency    int32
	Memo       *string
	DueDate    pgtype.Timestamptz
	CreatedAt  pgtype.Timestamptz
	UpdatedAt  pgtype.Timestamptz
}

func CreateTask(t *testing.T, pool *pgxpool.Pool, p TaskParams) repository.Task {
	t.Helper()
	if p.ID == uuid.Nil {
		id, err := uuid.NewV7()
		if err != nil {
			t.Fatalf("testfactory.CreateTask: uuid.NewV7: %v", err)
		}
		p.ID = id
	}
	if p.Title == "" {
		p.Title = "Test Task"
	}
	if p.Importance == 0 {
		p.Importance = 1
	}
	if p.Urgency == 0 {
		p.Urgency = 1
	}
	if !p.CreatedAt.Valid {
		p.CreatedAt = pgtype.Timestamptz{Time: time.Now(), Valid: true}
	}
	if !p.UpdatedAt.Valid {
		p.UpdatedAt = pgtype.Timestamptz{Time: time.Now(), Valid: true}
	}
	q := repository.New(pool)
	task, err := q.CreateTask(context.Background(), repository.CreateTaskParams{
		ID:         p.ID,
		UserID:     p.UserID,
		Title:      p.Title,
		StatusID:   p.StatusID,
		Importance: p.Importance,
		Urgency:    p.Urgency,
		Memo:       p.Memo,
		DueDate:    p.DueDate,
		CreatedAt:  p.CreatedAt,
		UpdatedAt:  p.UpdatedAt,
	})
	if err != nil {
		t.Fatalf("testfactory.CreateTask: %v", err)
	}
	return task
}
